core/shared/src/main/scala/cats/parse/Parser.scala

```diff
@@ -1775,32 +1775,31 @@ object Parser extends ParserInstances {
      * Merge CharIn bitsets
      */
     def mergeCharIn[A, P <: Parser[A]](ps: List[P]): List[P] = {
-      def loop(ps: List[P], front: List[(Int, BitSetUtil.Tpe)]): List[P] = {
+      @annotation.tailrec
+      def loop(ps: List[P], front: List[(Int, BitSetUtil.Tpe)], result: List[P]): List[P] = {
         @inline
         def frontRes: List[P] =
-          front match {
-            case Nil => Nil
-            case nel => Parser.charIn(BitSetUtil.union(nel)).asInstanceOf[P] :: Nil
-          }
+          if (front.isEmpty) Nil
+          else Parser.charIn(BitSetUtil.union(front)).asInstanceOf[P] :: Nil
 
         ps match {
-          case Nil => frontRes
+          case Nil => result ::: frontRes
           case AnyChar :: tail =>
             // AnyChar is bigger than all subsequent CharIn:
             // and any direct prefix CharIns
             val tail1 = tail.filterNot(_.isInstanceOf[CharIn])
-            AnyChar.asInstanceOf[P] :: tail1
+            result ::: AnyChar.asInstanceOf[P] :: tail1
           case CharIn(m, bs, _) :: tail =>
-            loop(tail, (m, bs) :: front)
+            loop(tail, (m, bs) :: front, result)
           case h :: tail =>
             // h is not an AnyChar or CharIn
             // we make our prefix frontRes
             // and resume working on the tail
-            frontRes ::: (h :: loop(tail, Nil))
+            loop(tail, Nil, result ::: frontRes ::: h :: Nil)
```

Comment:
Concat on the right is O(N) on a list so this is quadratic no? If we say, the list isn't long, that also means it won't blow the stack. So in the case this change helps us it also causes N^2 complexity I think.

Can we use a Chain to get fast concat and then convert to a List at the end? Or append on the left and reverse at the end?

What do you think? Does that sound right?

Reply:
Right, got to the same conclusion as well. Updated it with `Chain` as it seems to fit this better.